packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart

```diff
@@ -0,0 +1,1296 @@
+  // was tracked in [addAllowedPointer].
+  //
+  // This value defaults to an empty set.
+  //
+  // When the timer between two taps elapses, the recognizer loses the arena, the gesture is cancelled
+  // or the recognizer is disposed of then this value is reset.
+  Set<LogicalKeyboardKey> get keysPressedOnDown => _keysPressedOnDown ?? <LogicalKeyboardKey>{};
+
+  // The upper limit for the [consecutiveTapCount]. When this limit is reached
+  // all tap related state is reset and a new tap series is tracked.
+  //
+  // If this value is null, [consecutiveTapCount] can grow infinitely large.
+  int? get maxConsecutiveTap;
+
+  // The maximum distance in logical pixels the gesture is allowed to drift
+  // from the initial touch down position before the [consecutiveTapCount]
+  // and [keysPressedOnDown] are frozen and the remaining tracker state is
+  // reset. These values remain frozen until the next [PointerDownEvent] is
+  // tracked in [addAllowedPointer].
+  double? get slopTolerance;
+
+  // Private tap state tracked.
+  PointerDownEvent? _down;
+  PointerUpEvent? _up;
+  int _consecutiveTapCount = 0;
+  Set<LogicalKeyboardKey>? _keysPressedOnDown;
+
+  bool _wonArena = false;
+  OffsetPair? _originPosition;
+  int? _previousButtons;
+
+  // For timing taps.
+  Timer? _consecutiveTapTimer;
+  Offset? _lastTapOffset;
+
+  // When tracking a tap, the [consecutiveTapCount] is incremented if the given tap
+  // falls under the tolerance specifications and reset to 1 if not.
+  @override
+  void addAllowedPointer(PointerDownEvent event) {
+    super.addAllowedPointer(event);
+    if (maxConsecutiveTap == _consecutiveTapCount) {
+      _tapTrackerReset();
+    }
+    _up = null;
+    _wonArena = false;
+    if (_down != null && !_representsSameSeries(event)) {
+      // The given tap does not match the specifications of the series of taps being tracked,
+      // reset the tap count and related state.
+      _consecutiveTapCount = 1;
+    } else {
+      _consecutiveTapCount += 1;
+    }
+    _consecutiveTapTimerStop();
+    // `_down` must be assigned in this method instead of [handleEvent],
+    // because [acceptGesture] might be called before [handleEvent],
+    // which may rely on `_down` to initiate a callback.
+    _trackTrap(event);
+  }
+
+  @override
+  void acceptGesture(int pointer) {
+    _wonArena = true;
+    if (_up != null && _down != null) {
+      _consecutiveTapTimerStop();
+      _consecutiveTapTimerStart();
+      _wonArena = false;
+    }
+  }
+
+  @override
+  void handleEvent(PointerEvent event) {
+    if (event is PointerMoveEvent) {
+      final bool isSlopPastTolerance = slopTolerance != null && _getGlobalDistance(event, _originPosition) > slopTolerance!;
+
+      if (isSlopPastTolerance) {
+        _consecutiveTapTimerStop();
+        _previousButtons = null;
+        _lastTapOffset = null;
+      }
+    } else if (event is PointerUpEvent) {
+      _up = event;
+      if (_wonArena && _down != null) {
```

Comment:
This can happen if we haven't sent a tap down (a tap down is sent when the recognizer wins the arena or the `kPressTimeout` elapses) before arrive at a `PointerUpEvent` in `handleEvent`. This can happen when a recognizer is competing with this one because this recognizer will only declare self-victory if:

1. It is the last recognizer in the arena
2. The `kPressTimeout` has elapsed and the consecutive tap count is greater than 1.
3. The pointer has moved a sufficient global distance to be considered a drag (moved past the panSlop).

We only want to start the timer on tap up when we are sure a tap down has been sent.